feat(apig/env): add a resources to manage the environment for the shared APIG

Zippo-Wang opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:

Add a resources to manage the environment for the shared APIG.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->

```release-note
NONE
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
=== RUN   TestAccApiGatewayEnvironment_basic
=== PAUSE TestAccApiGatewayEnvironment_basic
=== CONT  TestAccApiGatewayEnvironment_basic
--- PASS: TestAccApiGatewayEnvironment_basic (26.17s)
PASS
```

chengxiangdong commented on docs/resources/api_gateway_environment.md:
```diff
@@ -0,0 +1,47 @@
+
+# huaweicloud_api_gateway_environment
+
+Manages an APIG environment resource within HuaweiCloud.
```
```suggestion
Manages a shared APIG environment resource within HuaweiCloud.
```

chengxiangdong commented on docs/resources/api_gateway_environment.md:
```diff
@@ -0,0 +1,47 @@
+  The name must start with a letter.
+
+* `description` - (Optional, String) Specifies the environment description.
+  The value can contain a maximum of `255` characters, and the angle brackets (< and >) are not allowed.
```
```suggestion
  The value can contain a maximum of `255` characters.
```

chengxiangdong commented on docs/resources/api_gateway_environment.md:
```diff
@@ -0,0 +1,47 @@
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The ID of the dedicated environment.
```
```suggestion
* `id` - The environment ID.
```

chengxiangdong commented on huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go:
```diff
@@ -0,0 +1,97 @@
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("[DEBUG] envs is : %s", envs)
```
```suggestion
	log.Printf("[DEBUG] List of shared APIG environments: %#v", envs)
```
HTTP request will print request, response and header details, usually we don't print the result.

chengxiangdong commented on docs/resources/api_gateway_environment.md:
```diff
@@ -0,0 +1,47 @@
+
+* `id` - The ID of the dedicated environment.
+
+* `created_time` - The time when the environment was created.
```
```suggestion
* `create_time` - The time when the environment was created.
```

chengxiangdong commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,149 @@
+
+	n, err := GetEnvironment(apigClient, d.Get("name").(string), d.Id())
+	if err != nil {
+		return diag.Errorf("the environment is not exists: %s", err)
```
```suggestion
		return common.CheckDeletedDiag(d, err, "error retrieving shared APIG environment")
```

chengxiangdong commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,149 @@
+
+		_, err = environments.Update(apigClient, d.Id(), updateOpts)
+		if err != nil {
+			return diag.Errorf("error updating API gateway shared environment: %s", err)
```
```suggestion
			return diag.Errorf("error updating shared API gateway environment: %s", err)
```

chengxiangdong commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,149 @@
+
+	err = environments.Delete(apigClient, d.Id())
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error deleting API gateway shared environment")
```
```suggestion
		return common.CheckDeletedDiag(d, err, "error deleting shared API gateway environment")
```

chengxiangdong commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,149 @@
+	region := cfg.GetRegion(d)
+	apigClient, err := cfg.ApiGatewayV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating API gateway client: %s", err)
```
```suggestion
		return diag.Errorf("error creating shared API gateway client (v1) : %s", err)
```
Others also recommend printing this way.

chengxiangdong commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,149 @@
+
+	rst, err := environments.Create(apigClient, createOpts)
+	if err != nil {
+		return diag.Errorf("error creating API gateway shared environment: %s", err)
```
```suggestion
		return diag.Errorf("error creating shared API gateway environment: %s", err)
```

Lance52259 commented on docs/resources/api_gateway_environment.md:
```diff
@@ -0,0 +1,47 @@
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region where the instance is located.
```
```suggestion
* `region` - (Optional, String, ForceNew) Specifies the region where the environment is located.
```

Zippo-Wang replied:
I have fixed this problem

Lance52259 commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,147 @@
+	return resourceAPIGatewayEnvironmentRead(ctx, d, meta)
+}
+
+func GetEnvironment(client *golangsdk.ServiceClient, envName, envID string) (*environments.Environment, error) {
```
```suggestion
func getEnvironment(client *golangsdk.ServiceClient, envName, envID string) (*environments.Environment, error) {
```

Zippo-Wang replied:
I have fixed this problem😊

Lance52259 commented:
/lgtm

chengxiangdong commented:
/lgtm

ShiChangkuo commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,147 @@
+		Name:        d.Get("name").(string),
+		Description: d.Get("description").(string),
+	}
+	log.Printf("[DEBUG] Create API gateway shared options: %s", utils.MarshalValue(createOpts))
```
```suggestion
	log.Printf("[DEBUG] Create API gateway shared options: %v", createOpts)
```

Zippo-Wang replied:
> I wrote this like you before, but Xiangdong suggest me use this way to DEBUG. Can both methods be used?

I wrote this like you before, but Xiangdong suggest me use this way to DEBUG. Can both methods be used?

Zippo-Wang replied:
> 
I have fixed this problem, thanks for your valuable suggestions

ShiChangkuo commented on huaweicloud/services/apigateway/resource_huaweicloud_api_gateway_environment.go:
```diff
@@ -0,0 +1,147 @@
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"create_time": {
```
please define it as **created_at**

Zippo-Wang replied:
> please define it as **created_at**

I use "create_time" due to it is written in the API document. Do we have to ues "create_at"?

Zippo-Wang replied:
> please define it as **created_at**

I have fixed this problem, thanks for your valuable suggestions

ShiChangkuo commented on huaweicloud/services/acceptance/apigateway/resource_huaweicloud_api_gateway_environment_test.go:
```diff
@@ -0,0 +1,97 @@
+		return nil, fmt.Errorf("error creating APIG client %s", err)
+	}
+
+	envs, err := environments.List(client, environments.ListOpts{})
```
it's better to filter environments by `name`.

Zippo-Wang replied:
> it's better to filter environments by `name`.

The id and name of environments are both the unique, so both of these can be used to determine an environment

Zippo-Wang replied:
> it's better to filter environments by `name`.

I have fixed this problem, thanks for your valuable suggestions

Zippo-Wang committed:
feat(apig/env): support shared environment resource
thub.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/apigateway"
